src/current/v24.3/troubleshoot-lock-contention.md

```diff
@@ -0,0 +1,395 @@
+- Historical queries have the best possible performance, since they are served by the nearest replica.
+
+### &quot;fail fast&quot; method
+
+In this tutorial's [Example 2](#example-2), Transaction 6 uses a &quot;fail fast&quot; method:
+
+~~~ sql
+BEGIN;
+SELECT * FROM t2 WHERE k=4 FOR UPDATE NOWAIT; -- fail fast write
+UPDATE t2 SET v=4034 WHERE k=4;
+COMMIT;
+~~~
+
+Consider the following when using a ["fail fast" method]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies):
+
+- &quot;Fail fast&quot; could be a reasonable protective measure in the application to handle "hot update key" situations, for example, when an application needs to be able to handle an arbitrary large surge of updates on the same key.
+- The most direct method of "failing fast" is using pessimistic locking with [SELECT FOR UPDATE … NOWAIT]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies). It can reduce or prevent failures late in a transaction's life (e.g. at the `COMMIT` time), by returning an error early in a contention situation if a row cannot be locked immediately.
+- A more &quot;buffered fail fast&quot; approach would be to control the maximum length of a lock wait-queue that requests are willing to enter and wait in, with the cluster setting  [`kv.lock_table.maximum_lock_wait_queue_length`](https://github.com/cockroachdb/cockroach/pull/66146) (default: `0`). It can provide some level of quality-of-service with a response time predictability in a severe per-key contention. If set to a non-zero value and an existing lock wait-queue is already equal to or exceeding this length, requests will be rejected eagerly instead of entering the queue and waiting.
+
+### Initial Data for Example 2
+
+In any of the SQL shells, create a second table and insert some data:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+DROP TABLE IF EXISTS t2;
+CREATE TABLE t2 (k INT PRIMARY KEY, v INT);
+INSERT INTO t2 VALUES (4,4), (5,5), (6,6);
+~~~
+
+### Example 2
+
+In this example, Transaction 4 is a write that does not block either Transaction 5, a read, or Transaction 6, a write. Transaction 4 locks key `k=4`. When Transaction 5 tries to read key `k=4`, it does not experience lock contention because it does not have to wait for the lock on the key to be released. Transaction 5 uses [`AS OF SYSTEM TIME`]({% link {{ page.version.version }}/as-of-system-time.md %}) to do a historical read. When Transaction 6 executes the [`SELECT ... FOR UPDATE NOWAIT`]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies) on key `k=4`, an error is returned since the key `k=4` cannot be locked immediately. In other words, Transaction 6 "fails fast". It does not even attempt to do an `UPDATE` write to key `k=4`, so it does not experience lock contention.
```

Comment:
suggest adding more overview here before jumping right into the gory details, something like "in this example, we will show how to do ${HIGH LEVEL DESCRIPTION OF THE EXAMPLE}.  Transaction 4 is a write that does not block ..."

Reply:
added more overview